src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,465 @@
+	}
+
+	public OffsetManager getOffsetManager() {
+		return offsetManager;
+	}
+
+	public void setOffsetManager(OffsetManager offsetManager) {
+		this.offsetManager = offsetManager;
+	}
+
+	public MessageListener getMessageListener() {
+		return messageListener;
+	}
+
+	public void setMessageListener(MessageListener messageListener) {
+		this.messageListener = messageListener;
+	}
+
+	public int getConcurrency() {
+		return concurrency;
+	}
+
+	/**
+	 * The maximum number of concurrent {@link MessageListener} running. Messages from within the same
+	 * partitions will be processed sequentially.
+	 */
+	public void setConcurrency(int concurrency) {
+		this.concurrency = concurrency;
+	}
+
+	public Executor getFetchTaskExecutor() {
+		return fetchTaskExecutor;
+	}
+
+	/**
+	 * The task executor for fetch operations
+	 */
+	public void setFetchTaskExecutor(Executor fetchTaskExecutor) {
+		this.fetchTaskExecutor = fetchTaskExecutor;
+	}
+
+	/**
+	 * The task executor for leader and offset updates.
+	 */
+	public Executor getAdminTaskExecutor() {
+		return adminTaskExecutor;
+	}
+
+	public void setAdminTaskExecutor(Executor adminTaskExecutor) {
+		this.adminTaskExecutor = adminTaskExecutor;
+	}
+
+	/**
+	 * The maximum amount of data (in bytes) that pollers will fetch in one round
+	 */
+	public int getMaxFetch() {
+		return maxFetch;
+	}
+
+	public void setMaxFetch(int maxFetch) {
+		this.maxFetch = maxFetch;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		synchronized (lifecycleMonitor) {
+			if (running) {
+				this.running = false;
+				this.messageDispatcher.stop();
+			}
+		}
+	}
+
+	@Override
+	public void start() {
+		synchronized (lifecycleMonitor) {
+			if (!running) {
+				this.running = true;
+				if (this.offsetManager == null) {
+					this.offsetManager = new MetadataStoreOffsetManager(kafkaTemplate.getConnectionFactory());
+				}
+				// initialize the fetch offset table - defer to OffsetManager for retrieving them
+				ImmutableList<Partition> partitionsAsList = Lists.immutable.with(partitions);
+				this.fetchOffsets = new ConcurrentHashMap<Partition, Long>(partitionsAsList.toMap(passThru, getOffset));
+				this.messageDispatcher = new ConcurrentMessageListenerDispatcher(messageListener, Arrays.asList(partitions), concurrency, offsetManager);
+				this.messageDispatcher.start();
+				partitionsByBrokerMap.putAll(partitionsAsList.groupBy(getLeader));
+				if (fetchTaskExecutor == null) {
+					fetchTaskExecutor = Executors.newFixedThreadPool(partitionsByBrokerMap.size());
+				}
+				partitionsByBrokerMap.forEachKey(launchFetchTask);
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		this.stop(null);
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return 0;
```

Comment:
We can add a setter for the phase - or perhaps return `Integer.MAXVALUE`, as other implementations do. I don't think that the Phase is relevant, but it's required because of `SmartLifecycle`.